File: javascript/packages/linter/src/rules/html-no-nested-forms.ts
import { BaseRuleVisitor } from "./rule-utils.js"
import { getHelpersForTag, getTagLocalName, isERBOutputNode, isHTMLOpenTagNode, PrismVisitor } from "@herb-tools/core"
import { ParserRule } from "../types.js"

import type { UnboundLintOffense, LintContext, FullRuleConfig } from "../types.js"
import type { HTMLElementNode, ERBBlockNode, ERBContentNode, ParseResult, ParserOptions, PrismNode, Location } from "@herb-tools/core"

// Supported helpers are transformed into `<form>` elements by the parser and
// caught by the element branch, so only unsupported ones need Prism detection.
const FORM_HELPERS = new Set(
  getHelpersForTag("form")
    .filter(helper => !helper.supported)
    .flatMap(helper => [helper.name, ...helper.aliases])
)

class FormHelperCallCollector extends PrismVisitor {
  public helperName: string | null = null

  visitCallNode(node: PrismNode): void {
    if (this.helperName) return

    if (!node.receiver && FORM_HELPERS.has(node.name)) {
      this.helperName = node.name
      return
    }

    this.visitChildNodes(node)
  }
}

class NestedFormVisitor extends BaseRuleVisitor {
  private formDepth = 0

  visitHTMLElementNode(node: HTMLElementNode): void {
    if (getTagLocalName(node) !== "form") {
      super.visitHTMLElementNode(node)
      return
    }

    if (this.formDepth > 0) {
      this.addOffense(
        "Nested `<form>` elements are not allowed. Move the inner `<form>` outside of the enclosing form, or associate its controls using the `form` attribute.",
        this.elementLocation(node),
      )
    }

    this.formDepth++
    super.visitHTMLElementNode(node)
    this.formDepth--
  }

  visitERBBlockNode(node: ERBBlockNode): void {
    const helperName = this.formHelperName(node)

    if (!helperName) {
      super.visitERBBlockNode(node)
      return
    }

    this.checkNestedHelper(helperName, node.location)

    this.formDepth++
    super.visitERBBlockNode(node)
    this.formDepth--
  }

  visitERBContentNode(node: ERBContentNode): void {
    const helperName = this.formHelperName(node)

    if (helperName) {
      this.checkNestedHelper(helperName, node.location)
    }

    super.visitERBContentNode(node)
  }

  private checkNestedHelper(helperName: string, location: Location): void {
    if (this.formDepth === 0) return

    this.addOffense(
      `\`${helperName}\` renders its own \`<form>\` element and cannot be nested inside another form. Move it outside of the enclosing form.`,
      location,
    )
  }

  private formHelperName(node: ERBBlockNode | ERBContentNode): string | null {
    if (!isERBOutputNode(node)) return null

    const prismNode = node.prismNode
    if (!prismNode) return null

    const collector = new FormHelperCallCollector()
    collector.visit(prismNode)

    return collector.helperName
  }

  private elementLocation(node: HTMLElementNode): Location {
    if (isHTMLOpenTagNode(node.open_tag) && node.open_tag.tag_name) {
      return node.open_tag.tag_name.location
    }

    return node.location
  }
}

export class HTMLNoNestedFormsRule extends ParserRule {
  static ruleName = "html-no-nested-forms"
  static introducedIn = this.version("unreleased")

  get defaultConfig(): FullRuleConfig {
    return {
      enabled: true,
      severity: "error"
    }
  }

  get parserOptions(): Partial<ParserOptions> {
    return {
      action_view_helpers: true,
      prism_nodes: true,
    }
  }

  check(result: ParseResult, context?: Partial<LintContext>): UnboundLintOffense[] {
    const visitor = new NestedFormVisitor(this.ruleName, context)

    visitor.visit(result.value)

    return visitor.offenses
  }
}

File: rust/herb-linter/src/rules/html_no_nested_forms.rs
use std::collections::HashSet;

use crate::offense::UnboundOffense;
use crate::rule::{LintContext, ParserRule, Rule};
use crate::utils::prism_utils::walk_prism;
use crate::utils::tag_utils::{element_tag_name_location, get_tag_local_name, is_erb_output_node};

use herb::nodes::{ERBBlockNode, ERBContentNode, HTMLElementNode};
use herb::prism::PrismNode;
use herb::Location;
use herb::ParseResult;
use herb::Visitor;
use herb_config::{Severity, SeverityConfig};

pub struct HTMLNoNestedFormsRule;

impl Rule for HTMLNoNestedFormsRule {
  fn name(&self) -> &'static str {
    "html-no-nested-forms"
  }

  fn introduced_in(&self) -> Option<&'static str> {
    Some("unreleased")
  }

  fn default_severity(&self) -> SeverityConfig {
    SeverityConfig::Severity(Severity::Error)
  }

  fn parser_options(&self) -> herb::ParserOptions {
    herb::ParserOptions {
      action_view_helpers: true,
      prism_nodes: true,
      ..crate::rule::default_linter_parser_options()
    }
  }
}

fn form_helper_names() -> HashSet<&'static str> {
  let mut names = HashSet::new();

  for entry in herb::action_view_helpers::helpers_for_tag("form") {
    if entry.supported {
      continue;
    }

    names.insert(entry.name);
    names.extend(entry.aliases.iter().copied());
  }

  names
}

struct NestedFormVisitor {
  rule_name: &'static str,
  offenses: Vec<UnboundOffense>,
  form_depth: usize,
  form_helpers: HashSet<&'static str>,
}

impl NestedFormVisitor {
  fn add_offense(&mut self, message: impl Into<String>, location: Location) {
    self.offenses.push(UnboundOffense::new(self.rule_name, message, location));
  }

  fn form_helper_name(&self, prism_node: Option<&PrismNode>) -> Option<&'static str> {
    let prism_node = prism_node?;
    let mut found = None;

    walk_prism(prism_node, &mut |node| {
      if found.is_some() {
        return false;
      }

      if node.is("CallNode") && node.receiver().is_none() {
        if let Some(name) = node.name.as_deref() {
          if let Some(helper) = self.form_helpers.get(name) {
            found = Some(*helper);
            return false;
          }
        }
      }

      true
    });

    found
  }

  fn check_nested_helper(&mut self, helper_name: &str, location: Location) {
    if self.form_depth == 0 {
      return;
    }

    self.add_offense(
      format!("`{helper_name}` renders its own `<form>` element and cannot be nested inside another `<form>`. Move it outside of the enclosing `<form>`."),
      location,
    );
  }
}

impl Visitor for NestedFormVisitor {
  fn visit_html_element_node(&mut self, node: &HTMLElementNode) {
    if get_tag_local_name(node).as_deref() != Some("form") {
      self.walk_html_element_node(node);
      return;
    }

    if self.form_depth > 0 {
      let location = element_tag_name_location(node);

      self.add_offense(
        "Nested `<form>` elements are not allowed. Move the inner `<form>` outside of the enclosing `<form>`, or associate its controls using the `form` attribute.",
        location,
      );
    }

    self.form_depth += 1;
    self.walk_html_element_node(node);
    self.form_depth -= 1;
  }

  fn visit_erb_block_node(&mut self, node: &ERBBlockNode) {
    let helper_name = if is_erb_output_node_block(node) {
      self.form_helper_name(node.prism())
    } else {
      None
    };

    let helper_name = match helper_name {
      Some(name) => name,
      None => {
        self.walk_erb_block_node(node);
        return;
      }
    };

    self.check_nested_helper(helper_name, node.location.clone());

    self.form_depth += 1;
    self.walk_erb_block_node(node);
    self.form_depth -= 1;
  }

  fn visit_erb_content_node(&mut self, node: &ERBContentNode) {
    if is_erb_output_node(node) {
      if let Some(helper_name) = self.form_helper_name(node.prism()) {
        self.check_nested_helper(helper_name, node.location.clone());
      }
    }

    self.walk_erb_content_node(node);
  }
}

fn is_erb_output_node_block(node: &ERBBlockNode) -> bool {
  node
    .tag_opening
    .as_ref()
    .map(|token| token.value == "<%=" || token.value == "<%==")
    .unwrap_or(false)
}

impl ParserRule for HTMLNoNestedFormsRule {
  fn check(&self, result: &ParseResult, _context: &LintContext) -> Vec<UnboundOffense> {
    let mut visitor = NestedFormVisitor {
      rule_name: self.name(),
      offenses: Vec::new(),
      form_depth: 0,
      form_helpers: form_helper_names(),
    };

    visitor.visit_document_node(&result.value);

    visitor.offenses
  }
}
